src/main/java/com/videoplaza/dataflow/pubsub/PubsubSourceTask.java

```diff
@@ -0,0 +1,318 @@
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static com.videoplaza.dataflow.pubsub.TimeUtils.msSince;
+import static com.videoplaza.dataflow.pubsub.TimeUtils.msTo;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * TODO report metrics
+ */
+public class PubsubSourceTask extends SourceTask {
+
+   private static final AtomicInteger TASK_COUNT = new AtomicInteger();
+
+   private final int id = TASK_COUNT.incrementAndGet();
+   private final Logger log = LoggerFactory.getLogger(PubsubSourceTask.class.getName() + "-" + id);
+
+   private final TaskMetrics metrics = new TaskMetrics();
+   private final Map<String, MessageInFlight> messages = new ConcurrentHashMap<>();
+   private final Set<PubsubMessage> toBePolled = ConcurrentHashMap.newKeySet();
+
+   private final AtomicBoolean stopping = new AtomicBoolean();
+   private final Lock pollCommitLock = new ReentrantLock();
+   private final Lock receiveLock = new ReentrantLock();
+   private final Condition recordsReceived = receiveLock.newCondition();
+
+   private volatile Subscriber subscriber;
+   private volatile PubsubSourceConnectorConfig config;
+   private volatile PubsubMessageConverter converter;
+   private volatile Sleeper sleeper;
+
+   @Override public String version() {
+      return Version.getVersion();
+   }
+
+   @Override public void start(Map<String, String> props) {
+      configure(props);
+      configure(new SimpleSleeper());
+      subscribe(newSubscriber());
+      log.info("Started");
+   }
+
+   PubsubSourceTask configure(Sleeper sleeper) {
+      this.sleeper = sleeper;
+      return this;
+   }
+
+   Subscriber newSubscriber() {
+
+      Subscriber subscriber = Subscriber.newBuilder(config.getProjectSubscription(), this::onPubsubMessageReceived)
+          .setFlowControlSettings(config.getFlowControlSettings())
+          .build();
+
+      subscriber.addListener(new Subscriber.Listener() {
+         public void failed(Subscriber.State from, Throwable failure) {
+            log.error("Failed {}.", from, failure);
+         }
+
+         @Override public void stopping(ApiService.State from) {
+            log.info("Stopping at {}. {}", from, metrics());
+         }
+
+         @Override public void terminated(ApiService.State from) {
+            log.info("Terminated at {}. {}", from, metrics());
+         }
+      }, Executors.newSingleThreadExecutor());
+      return subscriber;
+   }
+
+   PubsubSourceTask subscribe(Subscriber subscriber) {
+      this.subscriber = subscriber;
+      subscriber.startAsync();
+      return this;
+   }
+
+   PubsubSourceTask configure(Map<String, String> props) {
+      config = new PubsubSourceConnectorConfig(props);
+      converter = new PubsubMessageConverter(
+          config.getKeyAttribute(),
+          config.getTimestampAttribute(),
+          config.getSubscription(),
+          config.getTopic()
+      );
+      return this;
+   }
+
+   /**
```

Comment:
Do you think this partial deduplication worth the effort? If you think so, leave it as is. Otherwise remove the dups checking/book keeping

Reply:
I appears to be helpful on occasions, so that should stay IMO.